src/transformation/visitors/class/index.ts

```diff
@@ -349,3 +353,15 @@ export const transformSuperExpression: FunctionVisitor<ts.SuperExpression> = (ex
 };
 
 export const transformThisExpression: FunctionVisitor<ts.ThisExpression> = node => createSelfIdentifier(node);
+
+function getReflectionClassName(declaration: ts.ClassLikeDeclaration): string {
+    if (declaration.name) {
+        return declaration.name.text;
+    } else if (ts.isVariableDeclaration(declaration.parent) && ts.isIdentifier(declaration.parent.name)) {
```

Comment:
Have you seen https://github.com/TypeScriptToLua/TypeScriptToLua/issues/584#issuecomment-495465046? I'm okay with leaving it as it is for now, but note that it wouldn't cover a lot of cases

Reply:
Since we have decided not to use type information for that, here's an implementation that should specifically handle #584

```ts
function getReflectionClassName(
    context: TransformationContext,
    declaration: ts.ClassLikeDeclaration,
    className: lua.Identifier
): lua.Expression {
    if (declaration.name) {
        return lua.createStringLiteral(declaration.name.text);
    } else if (ts.isVariableDeclaration(declaration.parent) && ts.isIdentifier(declaration.parent.name)) {
        return lua.createStringLiteral(declaration.parent.name.text);
    } else if (hasDefaultExportModifier(declaration)) {
        return lua.createStringLiteral("default");
    }

    if (getExtendedNode(context, declaration)) {
        return lua.createTableIndexExpression(
            lua.createTableIndexExpression(className, lua.createStringLiteral("____super")),
            lua.createStringLiteral("name")
        );
    }

    return lua.createStringLiteral("");
}
```